lib/network_monitor.c

```diff
@@ -758,6 +815,173 @@ static void insert_addr(struct mptcpd_nm *nm,
+{
+        struct rtattr *rta = rta_buf;
+
+        rta->rta_len = RTA_LENGTH(sizeof(uint32_t));
+        rta->rta_type = type;
+        *((uint32_t *) RTA_DATA(rta)) = value;
+
+        return RTA_SPACE(sizeof(uint32_t));
+}
+
+static size_t rta_add_address(void *rta_buf, unsigned short type,
+                              uint8_t family,
+                              const struct in6_addr *v6,
+                              const struct in_addr *v4)
+{
+        struct rtattr *rta = rta_buf;
+
+        rta->rta_type = type;
+
+        switch (family) {
+        case AF_INET6:
+                rta->rta_len = RTA_LENGTH(sizeof(struct in6_addr));
+                memcpy(RTA_DATA(rta), v6, sizeof(struct in6_addr));
+                return RTA_SPACE(sizeof(struct in6_addr));
+        case AF_INET:
+                rta->rta_len = RTA_LENGTH(sizeof(struct in_addr));
+                memcpy(RTA_DATA(rta), v4, sizeof(struct in_addr));
+                return RTA_SPACE(sizeof(struct in_addr));
+        }
+
+        return 0;
+}
+
+static void check_default_route(struct mptcpd_addr_info *ai);
+
+static void handle_rtm_timeout(struct l_timeout *timeout,
+                               void *user_data)
+{
+        struct mptcpd_addr_info *ai = user_data;
+
+        (void) timeout;
+        check_default_route(ai);
+}
+
+#define MPTCPD_MAX_ROUTE_CHECK          3
+
+static void schedule_route_check(struct mptcpd_addr_info *ai)
+{
+        if (ai->attempts++ > MPTCPD_MAX_ROUTE_CHECK) {
+                char str[INET6_ADDRSTRLEN];
+
+                l_debug("timeout while waiting for default route on address %s",
+                        mptcpd_addr_to_string(ai, str, INET6_ADDRSTRLEN));
+                mptcpd_addr_put(ai);
+                return;
+        }
+
+        if (!ai->timeout) {
+                ai->timeout = l_timeout_create_ms(1, handle_rtm_timeout, ai, NULL);
+                if (!ai->timeout) {
+                        l_error("can't harm route re-check timeout");
+                        mptcpd_addr_put(ai);
+                }
+        } else {
+                /* Use exponential backoff */
+                l_timeout_modify_ms(ai->timeout, 1 << ai->attempts);
+        }
+}
+
+static void handle_rtm_getroute(int error,
+                                uint16_t type,
+                                void const *data,
+                                uint32_t len,
+                                void *user_data)
+{
+        struct mptcpd_addr_info *ai = user_data;
+        char str[INET6_ADDRSTRLEN];
+        uint32_t ifindex = 0;
+        char *dst = NULL;
+
+        if (error != 0) {
+                l_info("can't resolved default route from %s interface %d: %d",
+                       mptcpd_addr_to_string(ai, str, INET6_ADDRSTRLEN),
+                       ai->index, error);
+                goto again;
+        }
+
+        (void) type;
+        assert(type == RTM_NEWROUTE);
```

Comment:
Should this be `RTM_GETROUTE`?
```suggestion
        assert(type == RTM_GETROUTE);
```

Reply:
No. The kernel replies to RTM_GETROUTE queries with RTM_NEWROUTE replies.

Reply:
I see.  Thanks for clarifying!